File: e-r-i-b/WebResources/web/scripts/pfp/financePlanCalculator.js
function FinancePlanCalculator(startCapitalAmount,quarterlyInvestAmount,
                                     startCapitalQuarterIncome,quarterlyInvestQuarterIncome,
                                     startFinancePlanDate)
{
    this.minNegativeAmount = -0.005;
    this.defaultQuarterCount = 10;
    this.QUARTERS_IN_YEAR = 4;

    //данные на основе которых производим вычисления
    this.startCapitalAmount = startCapitalAmount;
    this.quarterlyInvestAmount = quarterlyInvestAmount;
    this.startCapitalQuarterIncome = startCapitalQuarterIncome;
    this.quarterlyInvestQuarterIncome = quarterlyInvestQuarterIncome;
    this.startFinancePlanDate = startFinancePlanDate;
    this.endFinancePlanDate = startFinancePlanDate;
    this.lastInListTargetDate = null;

    //внутринние данные для рассчетов
    this.loanQuarterAmountList = null;
    this.loanList = [];
    this.targetQuarterAmountList = null;
    this.loanQuarterPaymentList = null;
    this.quarterCount = this.defaultQuarterCount;
    this.yearStep = 1;
    this.lastTarget = null; // цель "создание пенсионных накоплений"

    //выходные данные
    this.totalAmountList = [];

    this.negativeBalanceList = [];

    this.setStartCapitalAmount = function(startCapitalAmount)
    {
        this.startCapitalAmount = startCapitalAmount;
    };

    this.setQuarterlyInvestAmount = function(quarterlyInvestAmount)
    {
        this.quarterlyInvestAmount = quarterlyInvestAmount;
    };

    this.setStartCapitalQuarterIncome = function(startCapitalQuarterIncome)
    {
        this.startCapitalQuarterIncome = startCapitalQuarterIncome;
    };

    this.setQuarterlyInvestQuarterIncome = function(quarterlyInvestQuarterIncome)
    {
        this.quarterlyInvestQuarterIncome = quarterlyInvestQuarterIncome;
    };

    /**
     * Установить список кредитов клиента
     * Сам список кредитов нам не интересен, поэтому сразу рассчитаем суммы кредитов по кварталам
     * и суммы платежей по кредитам по кварталам
     * @param loanList - список кредитов
     */
    this.setLoanList = function(loanList)
    {
        this.loanQuarterAmountList = [];
        this.loanList = loanList;
        var personLoan = null;
        for(var key in loanList)
        {
            personLoan = loanList[key];
            var loanQuarterSumm = this.loanQuarterAmountList[personLoan.startDate];
            if(loanQuarterSumm)
                this.loanQuarterAmountList[personLoan.startDate] = loanQuarterSumm + personLoan.amount;
            else
                this.loanQuarterAmountList[personLoan.startDate] = personLoan.amount;
        }

        this.loanQuarterPaymentList = [];
        for(var key in loanList)
        {
            personLoan = loanList[key];
            var quarter = pfpMath.nextQuarter(personLoan.startDate);
            var endQuarter = pfpMath.nextQuarter(personLoan.endDate);
            while(quarter.getTime() < endQuarter.getTime())
            {
                var loanQuarterPayment = this.loanQuarterPaymentList[quarter];
                if(loanQuarterPayment)
                    this.loanQuarterPaymentList[quarter] = loanQuarterPayment + personLoan.quarterPayment;
                else
                    this.loanQuarterPaymentList[quarter] = personLoan.quarterPayment;
                quarter = pfpMath.nextQuarter(quarter);
            }
        }
    };

    /**
     * Установить список целей.
     * Сами цели нам не интересны, поэому хранить их здесь не будем.
     * Посчитаем суммы, необходимые для целей по кварталам, их и будем хранить в этом объекте.
     * @param targetList - список целей
     */
    this.setTargetList = function(targetList)
    {
        this.targetQuarterAmountList = [];
        this.lastTarget = null;
        var personTarget = null;
        for(var key in targetList)
        {
            personTarget = targetList[key];
            var targetQuarterAmount = this.targetQuarterAmountList[personTarget.plannedDate];
            if(targetQuarterAmount)
                this.targetQuarterAmountList[personTarget.plannedDate] = targetQuarterAmount + personTarget.amount;
            else
                this.targetQuarterAmountList[personTarget.plannedDate] = personTarget.amount;
            if(personTarget.veryLast)
                this.lastTarget = personTarget;
        }

        var lastTargetDate = null;
        for(key in targetList)
        {
            personTarget = targetList[key];
            if(lastTargetDate == null || lastTargetDate.getTime() < personTarget.plannedDate.getTime())
                lastTargetDate = personTarget.plannedDate;
        }
        if(lastTargetDate == null)
            this.quarterCount = this.defaultQuarterCount;
        else
        {
            this.lastInListTargetDate = lastTargetDate;
            this.quarterCount = this.calcQuarterCount(lastTargetDate);
        }
    };

    this.calcQuarterCount = function(lastTargetDate)
    {
        var quarterCount = pfpMath.getDiffInQuarters(this.startFinancePlanDate, lastTargetDate);
        if(quarterCount <= 4 * this.QUARTERS_IN_YEAR)
        {
            this.endFinancePlanDate = lastTargetDate;
            return quarterCount;
        }

        this.yearStep = 1;
        // Если период больше 26-ти лет, то шаг равен 16-ти кварталам (4 года).
        if(quarterCount > 26 * this.QUARTERS_IN_YEAR)
        {
            this.yearStep = 4;
        }
        // Если период больше 8-ми лет и меньше 26 лет, то шаг равен 8-ми кварталам (2 года).
        else if(quarterCount > 8 * this.QUARTERS_IN_YEAR)
        {
            this.yearStep = 2 ;
        }

        var lastYearQuarter = pfpMath.getLastQuarterForYear(lastTargetDate);
        if((lastYearQuarter.getFullYear() - this.startFinancePlanDate.getFullYear()) % this.yearStep != 0)
        {
            var yearCount = (lastYearQuarter.getFullYear() - this.startFinancePlanDate.getFullYear()) % this.yearStep;
            lastYearQuarter.setFullYear(lastYearQuarter.getFullYear() + this.yearStep - yearCount);
        }
        this.endFinancePlanDate = lastYearQuarter;

        return pfpMath.getDiffInQuarters(this.startFinancePlanDate, lastYearQuarter) + 1;
    };

    this.getYearStep = function()
    {
        return this.yearStep;
    };

    this.getYearCount = function()
    {
        return this.endFinancePlanDate.getFullYear() - this.startFinancePlanDate.getFullYear();
    };

    this.getQuartersInPeriod = function(fromYear, toYear)
    {
        var fromDate = this.startFinancePlanDate;
        var toDate = this.lastInListTargetDate;

        if (fromDate.getFullYear() < fromYear)
        {
            fromDate = new Date(fromYear, 0, 1);
        }
        if (toDate.getFullYear() > toYear)
        {
            toDate = new Date(toYear, 9, 1);
        }

        return pfpMath.getDiffInQuarters(fromDate, toDate);
    };

    this.getTotalAmount = function()
    {
        return this.totalAmountList;
    };

    this.getNegativeBalanceList = function()
    {
        return this.negativeBalanceList;
    };

    this.calculate = function()
    {
        this.totalAmountList = [];
        this.negativeBalanceList = [];
        var startCapitalAmountList = [];//стоимость портфеля стартовый капитал по кварталам
        var quarterlyInvestAmountList = []; //стоимость портфеля ежеквартальные вложения по кварталам

        var currentQuarterDate = this.startFinancePlanDate;

        startCapitalAmountList[currentQuarterDate] = this.startCapitalAmount;
        quarterlyInvestAmountList[currentQuarterDate] = 0;

        var totalAmount = {};
        totalAmount.date = currentQuarterDate;
        totalAmount.amount = this.startCapitalAmount;
        this.totalAmountList.push(totalAmount);

        var factor = 0;
        for(var i=1; i <= this.quarterCount; i++)
        {
            var lastTargetFactor = 0;
            var prevQuarterDate = currentQuarterDate;
            currentQuarterDate = pfpMath.nextQuarter(prevQuarterDate);

            var startCapitalPrevQuarterAmount = startCapitalAmountList[prevQuarterDate];
            var startCapitalQuarterAmount = startCapitalPrevQuarterAmount * (1 + this.startCapitalQuarterIncome);

            var loanQuarterPayment = this.loanQuarterPaymentList[currentQuarterDate];
            if(!loanQuarterPayment)
                loanQuarterPayment = 0;

            var quarterlyInvestPrevQuarterAmount = quarterlyInvestAmountList[prevQuarterDate];
            var currentQuarterlyInvestAmount = this.quarterlyInvestAmount - loanQuarterPayment;
            var quarterlyInvestQuarterAmount = currentQuarterlyInvestAmount + quarterlyInvestPrevQuarterAmount * (1 + this.quarterlyInvestQuarterIncome);
            //quarterlyInvestQuarterAmountList.put(i, currentQuarterlyInvestAmount);

            var currentQuarterLoanAmount = this.loanQuarterAmountList[currentQuarterDate];
            if (currentQuarterLoanAmount)
                factor = factor + currentQuarterLoanAmount;

            /*если цель не создание пенсионных накоплений, то на графике отображаем сумму за вычитом цели*/
            if (this.lastTarget == null || currentQuarterDate.getTime() < this.lastTarget.plannedDate.getTime())
            {
                var currentQuarterTargetAmount = this.targetQuarterAmountList[currentQuarterDate];
                if (currentQuarterTargetAmount)
                    factor = factor - currentQuarterTargetAmount;
            }
            else  /* иначе учитываем сумму цели только в списке кварталов с отрицательным балансом */
            {
                var currentQuarterTargetAmount = this.targetQuarterAmountList[currentQuarterDate];
                if (currentQuarterTargetAmount)
                    lastTargetFactor = currentQuarterTargetAmount;
            }

            if (factor < 0)
            {
                var totalSum = startCapitalQuarterAmount + quarterlyInvestQuarterAmount;
                var factorAbsoluteValue = Math.abs(factor);
                if (totalSum <= factorAbsoluteValue)
                {
                    startCapitalQuarterAmount = 0;
                    quarterlyInvestQuarterAmount = 0;
                    factor = factor + totalSum;
                }
                else
                {
                    var startCapitalRatio = startCapitalQuarterAmount / totalSum;
                    var quarterlyInvestRatio = quarterlyInvestQuarterAmount / totalSum;
                    startCapitalQuarterAmount = startCapitalQuarterAmount - startCapitalRatio * factorAbsoluteValue;
                    quarterlyInvestQuarterAmount = quarterlyInvestQuarterAmount - quarterlyInvestRatio * factorAbsoluteValue;
                    factor = 0;
                }
            }

            startCapitalAmountList[currentQuarterDate] = startCapitalQuarterAmount;
            quarterlyInvestAmountList[currentQuarterDate] =  quarterlyInvestQuarterAmount;

            totalAmount = {};
            totalAmount.date = currentQuarterDate;
            totalAmount.amount = startCapitalQuarterAmount + quarterlyInvestQuarterAmount + factor;
            this.totalAmountList.push(totalAmount);

            /* заполняем список теми кварталами, что дают отрицательный результат с учетом пенсионных накоплений */
            if(totalAmount.amount - lastTargetFactor <= this.minNegativeAmount)
            {
                this.negativeBalanceList.push({
                    date:totalAmount.date,
                    amount:Math.abs(totalAmount.amount - lastTargetFactor)
                });
            }
        }
    };
    
    this.canAddLoan = function(loan)
    {
        var currentLoanQuarterPaymentList = [];
        var loanList = [];
        //строим актуальный список кредитов
        for (var k in this.loanList)
            loanList[k] = this.loanList[k];
        var id = loan.id?loan.id:-1;
        loanList[id] = loan;
        //строим поквартальный список выплат по кредитам
        for(var key in loanList)
        {
            var personLoan = loanList[key];
            var quarter = pfpMath.nextQuarter(personLoan.startDate);
            var endQuarter = pfpMath.nextQuarter(personLoan.endDate);
            while(quarter.getTime() < endQuarter.getTime())
            {
                var loanQuarterPayment = currentLoanQuarterPaymentList[quarter];
                if(loanQuarterPayment)
                    currentLoanQuarterPaymentList[quarter] = loanQuarterPayment + personLoan.quarterPayment;
                else
                    currentLoanQuarterPaymentList[quarter] = personLoan.quarterPayment;
                quarter = pfpMath.nextQuarter(quarter);
            }
        }
        //проверяем, все ли во кварталы пользователь сможет оплатить кредиты
        for (var n in currentLoanQuarterPaymentList)
        {
            if (currentLoanQuarterPaymentList[n] > this.quarterlyInvestAmount)
                return false;
        }
        return true;
    }
}

File: e-r-i-b/BusinessPFP/src/com/rssl/phizic/operations/pfp/FinancePlanCalculator.java
package com.rssl.phizic.operations.pfp;

import com.rssl.phizic.business.dictionaries.pfp.common.PortfolioType;
import com.rssl.phizic.business.pfp.PersonLoan;
import com.rssl.phizic.business.pfp.PersonTarget;
import com.rssl.phizic.business.pfp.PersonalFinanceProfile;
import com.rssl.phizic.business.pfp.portfolio.PersonPortfolio;
import com.rssl.phizic.utils.DateHelper;

import java.math.BigDecimal;
import java.util.Calendar;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author lepihina
 * @ created 13.09.13
 * @ $Author$
 * @ $Revision$
 */
public class FinancePlanCalculator
{
	private BigDecimal MIN_NEGATIVE_AMOUNT = new BigDecimal(-0.01);
    private static final int DEFAULT_QUARTER_COUNT = 10;
	private static final int CALCULATION_SCALE = 10;
	private static final BigDecimal QUARTERS_IN_YEAR = BigDecimal.valueOf(4);
	private static final int MONTHS_IN_QUARTER = 3;
	private static final int QUARTER_IN_YEAR = 4;

    //данные на основе которых производим вычислени€
	private BigDecimal startCapitalAmount;
	private BigDecimal quarterlyInvestAmount;
	private BigDecimal startCapitalQuarterIncome;
	private BigDecimal quarterlyInvestQuarterIncome;
	private Calendar startFinancePlanDate;

	private Map<Integer, BigDecimal> loanQuarterAmountList = new HashMap<Integer, BigDecimal>();
	private Map<Integer, BigDecimal> loanQuarterPaymentList = new HashMap<Integer, BigDecimal>();
	private Map<Integer, BigDecimal> targetQuarterAmountList = new HashMap<Integer, BigDecimal>();

	private Map<Integer, BigDecimal> totalAmountList = new HashMap<Integer, BigDecimal>();
	private Map<Integer, BigDecimal> startCapitalAmountList = new HashMap<Integer, BigDecimal>(); //стоимость портфел€ стартовый капитал по кварталам
	private Map<Integer, BigDecimal> quarterlyInvestAmountList = new HashMap<Integer, BigDecimal>(); //стоимость портфел€ ежеквартальные вложени€ по кварталам
	private Map<Integer, BigDecimal> quarterlyInvestQuarterAmountList = new HashMap<Integer, BigDecimal>(); //ежеквартальные вложени€ по кварталам
	private Map<Integer, BigDecimal> negativeBalanceList = new HashMap<Integer, BigDecimal>();

	private int lastTargetQuarter = -1; // цель "создание пенсионных накоплений"
	private int quarterCount = DEFAULT_QUARTER_COUNT;

	/**
	 * @param personalFinanceProfile - анкета прохождени€ персонального финансового планировани€
	 */
	public FinancePlanCalculator(PersonalFinanceProfile personalFinanceProfile)
	{
		PersonPortfolio startCapital = personalFinanceProfile.getPortfolioByType(PortfolioType.START_CAPITAL);
		PersonPortfolio quarterlyInvest = personalFinanceProfile.getPortfolioByType(PortfolioType.QUARTERLY_INVEST);

		Calendar executionDate = personalFinanceProfile.getStartPlaningDate();
		this.startCapitalAmount = startCapital.getProductSum().getDecimal();
		this.quarterlyInvestAmount = quarterlyInvest.getProductSum().getDecimal();
		this.startCapitalQuarterIncome = startCapital.getIncome().divide(QUARTERS_IN_YEAR).divide(BigDecimal.valueOf(100));
		this.quarterlyInvestQuarterIncome = quarterlyInvest.getIncome().divide(QUARTERS_IN_YEAR).divide(BigDecimal.valueOf(100));
		this.startFinancePlanDate = executionDate;
		setQuartetCount(personalFinanceProfile);
		setTargetList(executionDate, personalFinanceProfile.getPersonTargets());
		setLoanList(executionDate, personalFinanceProfile.getPersonLoans());
		calculate();
	}

	/**
	 * @return суммы кредитов по кварталам
	 */
	public Map<Integer, BigDecimal> getLoanQuarterAmountList()
	{
		return loanQuarterAmountList;
	}

	/**
	 * @return суммы платежей по кредитам по кварталам
	 */
	public Map<Integer, BigDecimal> getLoanQuarterPaymentList()
	{
		return loanQuarterPaymentList;
	}

	public Map<Integer, BigDecimal> getTotalAmountList()
	{
		return totalAmountList;
	}

	public Map<Integer, BigDecimal> getStartCapitalAmountList()
	{
		return startCapitalAmountList;
	}

	public Map<Integer, BigDecimal> getQuarterlyInvestAmountList()
	{
		return quarterlyInvestAmountList;
	}

	public Map<Integer, BigDecimal> getQuarterlyInvestQuarterAmountList()
	{
		return quarterlyInvestQuarterAmountList;
	}

	public Map<Integer, BigDecimal> getNegativeBalanceList()
	{
		return negativeBalanceList;
	}

	public void setQuartetCount(PersonalFinanceProfile personalFinanceProfile)
	{
		Calendar endDate = (Calendar) personalFinanceProfile.getLastTargetDate().clone();
		quarterCount = DateHelper.getDiffInQuarters(startFinancePlanDate, endDate);
		if(quarterCount > 4*QUARTER_IN_YEAR)
		{
			// шаг отображени€ будет год (несколько лет), поэтому отображаем значени€ на последний квартал года
			int quarterDelta = QUARTER_IN_YEAR - DateHelper.getQuarter(endDate);
			quarterCount += quarterDelta;
		}
	}

	/**
	 * @return количество кварталов, на которое производ€тс€ расчеты
	 */
	public int getQuarterCount()
	{
		return quarterCount;
	}

	/**
	 * –асчитывает суммы целей по кварталам
	 * @param startDate - дата начала ѕ‘ѕ
	 * @param targetList - список целей клиента
	 */
	public void setTargetList(Calendar startDate, List<PersonTarget> targetList)
	{
        for(PersonTarget personTarget : targetList)
        {
	        Integer quarter = DateHelper.getDiffInQuarters(startDate, personTarget.getPlannedDate());
	        BigDecimal targetQuarterAmount = targetQuarterAmountList.get(quarter);
            if(targetQuarterAmount != null)
                targetQuarterAmountList.put(quarter, targetQuarterAmount.add(personTarget.getAmount().getDecimal()));
            else
	            targetQuarterAmountList.put(quarter, personTarget.getAmount().getDecimal());
	        if(personTarget.isVeryLast())
		        lastTargetQuarter = quarter;
        }
	}

	/**
	 * –асчитывает суммы кредитов по кварталам и суммы платежей по кредитам по кварталам
	 * @param startDate - дата начала ѕ‘ѕ
	 * @param loanList - список кредитов клиента
	 */
	public void setLoanList(Calendar startDate, List<PersonLoan> loanList)
	{
        for(PersonLoan personLoan : loanList)
        {
	        Integer quarter = DateHelper.getDiffInQuarters(startDate, personLoan.getStartDate());
	        BigDecimal loanQuarterSum = loanQuarterAmountList.get(quarter);
	        if(loanQuarterSum != null)
                loanQuarterAmountList.put(quarter, loanQuarterSum.add(personLoan.getAmount().getDecimal()));
            else
	            loanQuarterAmountList.put(quarter, personLoan.getAmount().getDecimal());
        }

		for(PersonLoan personLoan : loanList)
        {
	        Calendar quarter = DateHelper.getNextQuarter(personLoan.getStartDate());
	        Calendar endQuarter = DateHelper.getNextQuarter(personLoan.getEndDate());
	        while(quarter.before(endQuarter))
            {
	            Integer quarters = DateHelper.getDiffInQuarters(startDate, quarter);
	            BigDecimal loanQuarterPayment = loanQuarterPaymentList.get(quarters);
	            if(loanQuarterPayment != null)
		            loanQuarterPaymentList.put(quarters, loanQuarterPayment.add(PfpMathUtils.calcQuarterPayment(personLoan)));
	            else
		            loanQuarterPaymentList.put(quarters, PfpMathUtils.calcQuarterPayment(personLoan));
	            quarter = DateHelper.getNextQuarter(quarter);
            }
        }
	}

	/**
	 * ѕроизвести рассчет будующей стоимости портфелей
	 */
	public void calculate()
	{
		startCapitalAmountList.put(0, startCapitalAmount);
		quarterlyInvestAmountList.put(0, BigDecimal.ZERO);
		quarterlyInvestQuarterAmountList.put(0, BigDecimal.ZERO);
		totalAmountList.put(0, startCapitalAmount);
		BigDecimal factor = BigDecimal.ZERO;

        for(int i=1; i <= quarterCount; i++)
        {
	        BigDecimal startCapitalPrevQuarterAmount = startCapitalAmountList.get(i-1);
	        BigDecimal startCapitalQuarterAmount = startCapitalPrevQuarterAmount.multiply(startCapitalQuarterIncome.add(BigDecimal.ONE));

	        BigDecimal loanQuarterPayment = loanQuarterPaymentList.get(i);
	        if (loanQuarterPayment == null)
		        loanQuarterPayment = BigDecimal.ZERO;

	        BigDecimal quarterlyInvestPrevQuarterAmount = quarterlyInvestAmountList.get(i-1);
	        BigDecimal currentQuarterlyInvestAmount = quarterlyInvestAmount.subtract(loanQuarterPayment);
	        BigDecimal quarterlyInvestQuarterAmount = currentQuarterlyInvestAmount.add(quarterlyInvestPrevQuarterAmount.multiply(quarterlyInvestQuarterIncome.add(BigDecimal.ONE)));
	        quarterlyInvestQuarterAmountList.put(i, currentQuarterlyInvestAmount);

	        BigDecimal currentQuarterLoanAmount = loanQuarterAmountList.get(i);
	        if (currentQuarterLoanAmount != null)
		        factor = factor.add(currentQuarterLoanAmount);

	        if (lastTargetQuarter < 0 || i < lastTargetQuarter)
	        {
		        BigDecimal currentQuarterTargetAmount = targetQuarterAmountList.get(i);
		        if (currentQuarterTargetAmount != null)
			        factor = factor.subtract(currentQuarterTargetAmount);
	        }

			if (factor.compareTo(BigDecimal.ZERO) < 0)
			{

				BigDecimal totalSum = startCapitalQuarterAmount.add(quarterlyInvestQuarterAmount);
				BigDecimal factorAbsoluteValue = factor.abs();
				if (totalSum.compareTo(factorAbsoluteValue) <= 0)
				{
					startCapitalQuarterAmount = BigDecimal.ZERO;
					quarterlyInvestQuarterAmount = BigDecimal.ZERO;
					factor = factor.add(totalSum);
				}
				else
				{
					BigDecimal startCapitalRatio = startCapitalQuarterAmount.divide(totalSum, CALCULATION_SCALE, BigDecimal.ROUND_UP);
					BigDecimal quarterlyInvestRatio = quarterlyInvestQuarterAmount.divide(totalSum, CALCULATION_SCALE, BigDecimal.ROUND_UP);
					startCapitalQuarterAmount = startCapitalQuarterAmount.subtract(startCapitalRatio.multiply(factorAbsoluteValue));
					quarterlyInvestQuarterAmount = quarterlyInvestQuarterAmount.subtract(quarterlyInvestRatio.multiply(factorAbsoluteValue));
					factor = BigDecimal.ZERO;
				}
			}
	        startCapitalAmountList.put(i, startCapitalQuarterAmount);
            quarterlyInvestAmountList.put(i, quarterlyInvestQuarterAmount);
	        BigDecimal totalAmount = startCapitalQuarterAmount.add(quarterlyInvestQuarterAmount).add(factor);
            totalAmountList.put(i, totalAmount);

	        if (totalAmount.compareTo(MIN_NEGATIVE_AMOUNT) <= 0)
            {
                negativeBalanceList.put(i, totalAmount.abs());
            }
        }
	}
}
